hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlUnionTest.java

```diff
@@ -240,19 +242,122 @@ public void watermarkedStreamUnionAllTest() {
-        },
+                    while (iterator.hasNext(50, TimeUnit.MILLISECONDS) == HasNextResult.YES) {
+                        actualRows.add(new Row(iterator.next()));
+                    }
+                    if (actualRows.size() == 7 && expected.size() < 7) {
+                        // Because we drop late items after union all (thanks to UnionDropLateItemsTransposeRule),
+                        // it can happen that one of the late items makes it through. But both will never make it.
+                        expected.add(new Row(timestampTz(0L), 4));
+                    }
+                    assertThat(actualRows).containsExactlyInAnyOrderElementsOf(expected);
+                },
                 5);
     }
 
+    @Test
+    public void prunableUnionAllTest() {
+        instance().getConfig().addMapConfig(
+                new MapConfig("pMap1").setPartitioningAttributeConfigs(List.of(
+                        new PartitioningAttributeConfig("id")
+                )));
+        instance().getConfig().addMapConfig(
+                new MapConfig("pMap2").setPartitioningAttributeConfigs(List.of(
+                        new PartitioningAttributeConfig("id")
+                )));
+
+        IMap<Person, String> prunableMap1 = instance().getMap("pMap1");
+        IMap<Person, String> prunableMap2 = instance().getMap("pMap2");
+
+        createMapping("pMap1", Person.class, String.class);
+        createMapping("pMap2", Person.class, String.class);
+
+        for (int i = 0; i < 5; ++i) {
+            prunableMap1.put(new Person(i, "ABC" + i), "" + i);
+            prunableMap2.put(new Person(i, "ABC" + i), "" + i);
+        }
+
+        String sql = "(SELECT this FROM pMap1 WHERE id = 1) UNION ALL (SELECT this FROM pMap2 WHERE id = 1 OR id = 2)";
+        expected.add(new Row("1"));
+        expected.add(new Row("1"));
+        expected.add(new Row("2"));
+
+        assertRowsAnyOrder(sql, expected);
+    }
+
+    @Test
+    public void prunableUnionTest() {
+        instance().getConfig().addMapConfig(
+                new MapConfig("pMap1").setPartitioningAttributeConfigs(List.of(
+                        new PartitioningAttributeConfig("id")
+                )));
+        instance().getConfig().addMapConfig(
+                new MapConfig("pMap2").setPartitioningAttributeConfigs(List.of(
+                        new PartitioningAttributeConfig("id")
+                )));
+
+        IMap<Person, String> prunableMap1 = instance().getMap("pMap1");
+        IMap<Person, String> prunableMap2 = instance().getMap("pMap2");
+
+        createMapping("pMap1", Person.class, String.class);
+        createMapping("pMap2", Person.class, String.class);
+
+        for (int i = 0; i < 5; ++i) {
+            prunableMap1.put(new Person(i, "ABC" + i), "" + i);
+            prunableMap2.put(new Person(i, "ABC" + i), "" + i);
+        }
+
+        String sql = "(SELECT this FROM pMap1 WHERE id = 1) UNION (SELECT this FROM pMap2 WHERE id = 1)";
```

Comment:
does not use member pruning:
`Job config for 0a39-dea1-42c0-0001: JobConfig {name=null, processingGuarantee=NONE, snapshotIntervalMillis=10000, autoScaling=true, suspendOnFailure=false, splitBrainProtectionEnabled=false, enableMetrics=true, storeMetricsAfterJobCompletion=false, resourceConfigs={}, serializerConfigs={}, arguments={__sql.queryUnbounded=false, __sql.arguments=[], __sql.queryText=(SELECT this FROM pMap1 WHERE id = 1) UNION (SELECT this FROM pMap2 WHERE id = 1)}, classLoaderFactory=null, initialSnapshotName=null, maxProcessorAccumulatedRecords=-1, timeoutMillis=0}`
From https://hazelcast.atlassian.net/browse/HZ-2693?focusedCommentId=81303 I understand it should.

Reply:
physical execution plan uses separate `DISTINCT` step:
```
AggregateCombineByKeyPhysicalRel(group=[{0}])
  AggregateAccumulateByKeyPhysicalRel(group=[{0}])
    UnionPhysicalRel(all=[true])
      FullScanPhysicalRel(table=[[hazelcast, public, pMap1[projects=[$3], filter==($0, 1)]]], discriminator=[0])
      FullScanPhysicalRel(table=[[hazelcast, public, pMap2[projects=[$3], filter==($0, 1)]]], discriminator=[0])
```

Reply:
And it shouldn't, `Aggregation` is not prunable yet. It's a required test when we will implement `Aggregation`, because this case may be easily missed.

Reply:
https://hazelcast.atlassian.net/browse/HZ-2796

Reply:
there are many solution to that, for example adding a comment with explanation and link to the issue